src/DotLiquid/Util/NumericConverter.cs

```diff
@@ -0,0 +1,190 @@
+                return left;
+            if (leftTypes.Contains(right))
+                return right;
+            return rightTypes.First(t => leftTypes.Contains(t));
+        }
+
+        /// <summary>
+        /// Coerce an object into a numeric type.
+        /// </summary>
+        /// <param name="value">The string to coerce.</param>
+        /// <param name="formatProvider">The format provider for converting floating point numbers.</param>
+        /// <param name="defaultValue">The value to return if coercion fails.</param>
+        /// <returns>The coerced value as int, long, double or decimal type, or <paramref name="defaultValue"/> if coercion fails.</returns>
+        public static object CoerceToNumericType(object value, IFormatProvider formatProvider, object defaultValue)
+        {
+            if (NumericConverter.TryCoerceToNumericType(value, formatProvider, out object convertedValue))
+            {
+                return convertedValue;
+            }
+
+            return defaultValue;
+        }
+
+        /// <summary>
+        /// Try to coerce an object into a numeric type.
+        /// </summary>
+        /// <param name="value">The string to parse.</param>
+        /// <param name="formatProvider">The format provider for converting floating point numbers.</param>
+        /// <param name="convertedValue">The coerced value as int, long, double or decimal type, or null if parsing fails.</param>
+        /// <returns>true if parsing was successful; Otherwise, false.</returns>
+        public static bool TryCoerceToNumericType(object value, IFormatProvider formatProvider, out object convertedValue)
+        {
+            if (value != null)
+            {
+                if (NumericConverter.IsNumeric(value))
+                {
+                    convertedValue = value;
+                    return true;
+                }
+                else if (value is string stringValue)
+                {
+                    return NumericConverter.TryParseToNumericType(stringValue, formatProvider, out convertedValue);
+                }
+            }
+
+            convertedValue = null;
+            return false;
+        }
+
+        /// <summary>
+        /// Try to parse the string into a numeric type.
+        /// </summary>
+        /// <param name="value">The string to parse.</param>
+        /// <param name="formatProvider">The format provider for converting floating point numbers.</param>
+        /// <param name="convertedValue">The coerced value as int, long, double or decimal type, or null if parsing fails.</param>
+        /// <returns>true if parsing was successful; Otherwise, false.</returns>
+        public static bool TryParseToNumericType(string value, IFormatProvider formatProvider, out object convertedValue)
+        {
+            if (value != null)
+            {
+                if (int.TryParse(value, NumberStyles.Integer | NumberStyles.AllowThousands, formatProvider, out int intValue))

```

Comment:
This is a change from the Regex in that thousands is now allowed while previously it was not. An improvement, certainly, but should be documented.

Reply:
Added `<remarks>` to XML documentation.